File: frontend/src/types/task.ts
export interface Subtask {
  id: number;
  text: string;
  completed: boolean;
}

export interface Task {
  id: number;
  title: string;
  description: string;
  status: "PENDING" | "COMPLETED";
  priority: "LOW" | "MEDIUM" | "HIGH";
  category: string | null;
  dueDate: string | null;
  createdAt: string;
  updatedAt: string;
  subtasks: Subtask[];
}

export interface TaskCreateRequest {
  title: string;
  description?: string;
  priority?: "LOW" | "MEDIUM" | "HIGH";
  category?: string;
  dueDate?: string;
  subtasks?: string[];
}

export interface DashboardStats {
  total: number;
  pending: number;
  completed: number;
  highPriority: number;
  overdue: number;
}

export interface RegisterRequest {
  name: string;
  email: string;
  password: string;
  confirmPassword: string;
  securityQuestion: string;
  securityAnswer: string;
}

export interface LoginRequest {
  email: string;
  password: string;
}

export interface AuthResponse {
  token: string;
  name: string;
  email: string;
}

File: backend/src/main/java/com/taskmanager/entity/Task.java
package com.taskmanager.entity;

import jakarta.persistence.*;
import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "tasks")
public class Task {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private TaskStatus status = TaskStatus.PENDING;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private TaskPriority priority = TaskPriority.MEDIUM;

    private String category;

    private LocalDate dueDate;

    private LocalDateTime createdAt;

    private LocalDateTime updatedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    // CascadeType.ALL means: when you delete a task, delete its subtasks too
    // orphanRemoval means: if a subtask is removed from this list, delete it from DB
    @OneToMany(mappedBy = "task", cascade = CascadeType.ALL, orphanRemoval = true)
    private java.util.List<Subtask> subtasks = new java.util.ArrayList<>();


    public Task() {}

    public Task(String title, String description, TaskPriority priority,
                String category, LocalDate dueDate, User user) {
        this.title = title;
        this.description = description;
        this.priority = (priority != null) ? priority : TaskPriority.MEDIUM;
        this.status = TaskStatus.PENDING;
        this.category = category;
        this.dueDate = dueDate;
        this.user = user;
    }


    public Long getId() { return id; }
    public String getTitle() { return title; }
    public String getDescription() { return description; }
    public TaskStatus getStatus() { return status; }
    public TaskPriority getPriority() { return priority; }
    public String getCategory() { return category; }
    public LocalDate getDueDate() { return dueDate; }
    public LocalDateTime getCreatedAt() { return createdAt; }
    public LocalDateTime getUpdatedAt() { return updatedAt; }
    public User getUser() { return user; }
    public java.util.List<Subtask> getSubtasks() { return subtasks; }


    public void setTitle(String title) { this.title = title; }
    public void setDescription(String description) { this.description = description; }
    public void setStatus(TaskStatus status) { this.status = status; }
    public void setPriority(TaskPriority priority) { this.priority = priority; }
    public void setCategory(String category) { this.category = category; }
    public void setDueDate(LocalDate dueDate) { this.dueDate = dueDate; }
    public void setUser(User user) { this.user = user; }


    @PrePersist
    protected void onCreate() {
        this.createdAt = LocalDateTime.now();
        this.updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }
}